programs/bpf_loader/src/syscalls.rs

```diff
@@ -2107,136 +2305,1723 @@ fn call<'a>(
+                Err(SyscallError::BigNumberOpenSSLErrorUnknown)
             }
         }
     }
+}
 
-    #[test]
-    fn test_translate_type() {
-        // Pubkey
-        let pubkey = solana_sdk::pubkey::new_rand();
-        let addr = &pubkey as *const _ as u64;
-        let memory_mapping = MemoryMapping::new::<UserError>(
-            vec![MemoryRegion {
-                host_addr: addr,
-                vm_addr: 100,
-                len: std::mem::size_of::<Pubkey>() as u64,
-                vm_gap_shift: 63,
-                is_writable: false,
-            }],
-            &DEFAULT_CONFIG,
-        )
-        .unwrap();
-        let translated_pubkey =
-            translate_type::<Pubkey>(&memory_mapping, 100, &bpf_loader::id(), true).unwrap();
-        assert_eq!(pubkey, *translated_pubkey);
-
-        // Instruction
-        let instruction = Instruction::new_with_bincode(
-            solana_sdk::pubkey::new_rand(),
-            &"foobar",
-            vec![AccountMeta::new(solana_sdk::pubkey::new_rand(), false)],
+/// BIGNUM sol_bignum_from_u32
+struct SyscallBigNumFromU32<'a> {
+    cost: u64,
+    word_cost: u64,
+    word_div_cost: u64,
+    compute_meter: Rc<RefCell<dyn ComputeMeter>>,
+    loader_id: &'a Pubkey,
+}
+impl<'a> SyscallObject<BpfError> for SyscallBigNumFromU32<'a> {
+    fn call(
+        &mut self,
+        bn_ffi_out_addr: u64,
+        u32_val: u64,
+        _arg3: u64,
+        _arg4: u64,
+        _arg5: u64,
+        memory_mapping: &MemoryMapping,
+        result: &mut Result<u64, EbpfError<BpfError>>,
+    ) {
+        question_mark!(
+            self.compute_meter.consume(calc_bignum_cost!(self, 4f64)),
+            result
         );
-        let addr = &instruction as *const _ as u64;
-        let mut memory_mapping = MemoryMapping::new::<UserError>(
-            vec![MemoryRegion {
-                host_addr: addr,
-                vm_addr: 96,
-                len: std::mem::size_of::<Instruction>() as u64,
-                vm_gap_shift: 63,
-                is_writable: false,
-            }],
-            &DEFAULT_CONFIG,
-        )
-        .unwrap();
-        let translated_instruction =
-            translate_type::<Instruction>(&memory_mapping, 96, &bpf_loader::id(), true).unwrap();
-        assert_eq!(instruction, *translated_instruction);
-        memory_mapping.resize_region::<BpfError>(0, 1).unwrap();
-        assert!(
-            translate_type::<Instruction>(&memory_mapping, 100, &bpf_loader::id(), true).is_err()
+        // Get the output FfiBigNumber structure
+        let bn_syscall_result = question_mark!(
+            translate_type_mut::<FfiBigNumber>(
+                memory_mapping,
+                bn_ffi_out_addr,
+                self.loader_id,
+                true
+            ),
+            result
         );
+        // Process BigNum call and populate results
+        match BigNum::from_u32(u32_val as u32) {
+            Ok(bn) => bignum_to_ffibignumber(
```

Comment:
This first creates a `BigNum` and then an ffibignumber. Seems a lot of round trips.

If you want create a bignum from `u32` (or `u64`) you can just set the data[0] to that value, negative to false zero the  other data fields.

Is a syscall necessary for this?

Reply:
A u32 can be 1-4 bytes in length.

Reply:
Right I didn't realize the data section was a series of bytes, not `u64`.

Having said that, this function is simply `u32::to_be_bytes()` with some bounds checks.

Reply:
@seanyoung Right, the FfiBigNumber is just a convenience structure when calling a `syscall`. I will add the "in program" change for `from_u32` as well as a adding a `from_u64` which executes in program space only